# Wenpei Yan Final Project

## Introduction/Business Problem

New York City and Toronto are very diverse and are the financial capitals of their respective countries.  
In this project, I will compare compare the neighborhoods of the two cities and determine how similar or dissimilar they are.


## Data Description

I will use the data Neighborhood available on Foursquare.  
Neighborhood has a total of 5 boroughs and 306 neighborhoods. In order to segement the neighborhoods and explore them, I will essentially need a dataset that contains the 5 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood.  
I will explore, analyze the cluster the neighborhoods to get a clear image of the two cities.   
After the analysis, I will be able to tell how similar or different the two citis are, what may be the inner culture and attitude of the two countries, and how these cultures influence the action of them respectively.

## Methodology

In [1]:
import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim

import requests # library to handle requests
from pandas.io.json import json_normalize 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    ca-certificates-2019.6.16  |       hecc5488_0         145 KB  conda-forge
    altair-3.1.0               |           py36_0         724 KB  conda-forge
    certifi-2019.6.16          |           py36_1         149 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.2 MB

The following NEW packages will be 

In [2]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [4]:
import json 
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [5]:
neighborhoods_data = newyork_data['features']
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 
neighborhoods = pd.DataFrame(columns=column_names)

In [6]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [7]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


### Create Map

In [8]:
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [9]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

In [10]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7900869, -73.9598295.


In [11]:
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

### get neighborhood data

In [12]:
CLIENT_ID = 'MXQE1KN1ORGNBPLN03FEUOPV0SV24MS3F3NTZLNUD0XMOIGO'
CLIENT_SECRET = 'L2WE2PO5U52P50DGVTWO5QF5BBT0AHJTI5NJK0XHQWYODNNT'
VERSION = '20180605' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: MXQE1KN1ORGNBPLN03FEUOPV0SV24MS3F3NTZLNUD0XMOIGO
CLIENT_SECRET:L2WE2PO5U52P50DGVTWO5QF5BBT0AHJTI5NJK0XHQWYODNNT


In [13]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

In [14]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


In [15]:
LIMIT = 100 
radius = 500 # define radius


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

In [16]:
results = requests.get(url).json()

In [17]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [18]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name   categories        lat        lng
0       Arturo's  Pizza Place  40.874412 -73.910271
1    Bikram Yoga  Yoga Studio  40.876844 -73.906204
2  Tibbett Diner        Diner  40.880404 -73.908937
3      Starbucks  Coffee Shop  40.877531 -73.905582
4        Dunkin'   Donut Shop  40.877136 -73.906666

In [19]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

26 venues were returned by Foursquare.


In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:

manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )



Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


In [22]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3315, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066      Starbucks   
4  Marble Hill              40.876551               -73.91066        Dunkin'   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877531       -73.905582    Coffee Shop  
4       40.877136       -73.906666     Donut Shop

In [23]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                       99                      99     99   
Carnegie Hill                          100                     100    100   
Central Harlem                          42                      42     42   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                           100                     100    100   
Clinton                                100                     100    100   
East Harlem                             43                      43     43   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                               100                     100    100   
Greenwich Village                      100                     100    100   
Hamilton Heights                        61                      61     61   
Hudson Yards                            75                      75     75   
Inwood                                  57                      57     57   
Lenox Hill                             100                     100    100   
Lincoln Square                         100                     100    100   
Little Italy                           100                     100    100   
Lower East Side                         60                      60     60   
Manhattan Valley                        56                      56     56   
Manhattanville                          40                      40     40   
Marble Hill                             26                      26     26   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     40                      40     40   
Murray Hill                            100                     100    100   
Noho                                   100                     100    100   
Roosevelt Island                        29                      29     29   
Soho                                   100                     100    100   
Stuyvesant Town                         17                      17     17   
Sutton Place                           100                     100    100   
Tribeca                                100                     100    100   
Tudor City                              83                      83     83   
Turtle Bay                             100                     100    100   
Upper East Side                        100                     100    100   
Upper West Side                        100                     100    100   
Washington Heights                      87                      87     87   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City                99               99              99  
Carnegie Hill                   100              100             100  
Central Harlem                   42               42              42  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                    100              100             100  
Clinton                         100              100             100  
East Harlem                      43               43              43  
East Village                    100              100             100  
Finan

In [24]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 342 uniques categories.


In [25]:
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0                 0                       0            0           0   
1                 0                       0            0           0   
2                 0                       0            0           0   
3                 0                       0            0           0   
4                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  BBQ Joint  \
0                      0                    0              0          0   
1                      0                    0              0          0   
2                      0                    0              0          0   
3                      0                    0              0          0   
4                      0                    0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0           0       0     0    0               0   
1           0           0       0     0    0               0   
2           0           0       0     0    0               0   
3           0           0       0     0    0               0   
4           0           0       0     0    0               0   

   Basketball Court  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0                 0                0         0            0           0   
1                 0                0         0            0           0   
2                 0                0         0            0           0   
3                 0                0         0            0           0   
4                 0                0         0            0           0   

   Big Box Store  Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  \
0              0                         0          0           0       0   
1              0                         0          0           0       0   
2              0                         0          0           0       0   
3              0                         0          0           0       0   
4              0                         0          0           0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Boxing Gym  \
0           0              0          0         0           0   
1           0              0          0         0           0   
2           0              0          0         0           0   
3           0              0          0         0           0   
4           0              0          0         0           0   

   Brazilian Restaurant  Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  \
0                     0               0         

In [26]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000            0.00               0.00   
1         Carnegie Hill           0.000000            0.00               0.00   
2        Central Harlem           0.000000            0.00               0.00   
3               Chelsea           0.000000            0.00               0.00   
4             Chinatown           0.000000            0.00               0.00   
5          Civic Center           0.000000            0.00               0.00   
6               Clinton           0.000000            0.00               0.00   
7           East Harlem           0.000000            0.00               0.00   
8          East Village           0.000000            0.00               0.00   
9    Financial District           0.010000            0.00               0.00   
10             Flatiron           0.000000            0.00               0.00   
11             Gramercy           0.000000            0.00               0.00   
12    Greenwich Village           0.000000            0.00               0.00   
13     Hamilton Heights           0.000000            0.00               0.00   
14         Hudson Yards           0.000000            0.00               0.00   
15               Inwood           0.000000            0.00               0.00   
16           Lenox Hill           0.000000            0.00               0.01   
17       Lincoln Square           0.000000            0.00               0.00   
18         Little Italy           0.000000            0.00               0.00   
19      Lower East Side           0.000000            0.00               0.00   
20     Manhattan Valley           0.000000            0.00               0.00   
21       Manhattanville           0.000000            0.00               0.00   
22          Marble Hill           0.000000            0.00               0.00   
23              Midtown           0.000000            0.00               0.00   
24        Midtown South           0.000000            0.00               0.00   
25  Morningside Heights           0.000000            0.00               0.00   
26          Murray Hill           0.000000            0.00               0.00   
27                 Noho           0.000000            0.01               0.00   
28     Roosevelt Island           0.000000            0.00               0.00   
29                 Soho           0.010000            0.00               0.00   
30      Stuyvesant Town           0.000000            0.00               0.00   
31         Sutton Place           0.000000            0.01               0.00   
32              Tribeca           0.000000            0.00               0.00   
33           Tudor City           0.000000            0.00               0.00   
34           Turtle Bay           0.000000            0.00               0.00   
35      Upper East Side           0.000000            0.00               0.00   
36      Upper West Side           0.010000            0.00               0.00   
37   Washington Heights           0.011494            0.00               0.00   
38         West Village           0.010000            0.00               0.00   
39            Yorkville           0.000000            0.00               0.00   

    African Restaurant  American Restaurant  Antique Shop  Arcade  \
0             0.000000             0.010101          0.00    0.00   
1             0.000000             0.010000          0.00    0.00   
2             0.071429             0.047619          0.00    0.00   
3             0.000000             0.030000          0.01    0.00   
4             0.000000             0.040000          0.00    0.00   
5             0.000000             0.030000          0.01    0.00   
6             0.000000             0.040000          0.00    0.00   
7             0.000000             0.000000          0.00    0.00   
8             0.000000             0.020000          0.01    0.00   

In [27]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
           venue  freq
0           Park  0.07
1    Coffee Shop  0.06
2          Hotel  0.05
3  Memorial Site  0.04
4            Gym  0.04


----Carnegie Hill----
         venue  freq
0  Coffee Shop  0.07
1  Pizza Place  0.06
2          Bar  0.04
3         Café  0.04
4       Bakery  0.03


----Central Harlem----
                 venue  freq
0   African Restaurant  0.07
1  American Restaurant  0.05
2                  Bar  0.05
3           Public Art  0.05
4   Seafood Restaurant  0.05


----Chelsea----
                venue  freq
0         Coffee Shop  0.06
1  Italian Restaurant  0.05
2      Ice Cream Shop  0.05
3           Nightclub  0.04
4              Bakery  0.04


----Chinatown----
                   venue  freq
0     Chinese Restaurant  0.09
1           Cocktail Bar  0.05
2  Vietnamese Restaurant  0.04
3    American Restaurant  0.04
4     Salon / Barbershop  0.04


----Civic Center----
                  venue  freq
0  Gym / Fitness Center  0.05
1    Italian

In [28]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [29]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park           Coffee Shop   
1      Carnegie Hill           Coffee Shop           Pizza Place   
2     Central Harlem    African Restaurant    Seafood Restaurant   
3            Chelsea           Coffee Shop        Ice Cream Shop   
4          Chinatown    Chinese Restaurant          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                 Hotel         Memorial Site                    Gym   
1                  Café                   Bar            Yoga Studio   
2   American Restaurant     French Restaurant             Public Art   
3    Italian Restaurant             Nightclub                 Bakery   
4    Salon / Barbershop   American Restaurant  Vietnamese Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Boat or Ferry             Wine Shop    Italian Restaurant   
1             Bookstore   Japanese Restaurant         Grocery Store   
2                   Bar    Chinese Restaurant          Dessert Shop   
3               Theater                 Hotel   American Restaurant   
4        Ice Cream Shop       Bubble Tea Shop                   Spa   

  9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant           Burger Joint  
1                Bakery                    Gym  
2              Beer Bar                Library  
3    Seafood Restaurant            Art Gallery  
4   Dumpling Restaurant                 Bakery

In [30]:
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 1, 1, 2, 1, 2, 2, 0, 1, 1], dtype=int32)

In [31]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               4   
1  Manhattan           Chinatown  40.715618 -73.994279               1   
2  Manhattan  Washington Heights  40.851903 -73.936900               0   
3  Manhattan              Inwood  40.867684 -73.921210               0   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Sandwich Place           Coffee Shop        Discount Store   
1    Chinese Restaurant          Cocktail Bar    Salon / Barbershop   
2                  Café                Bakery     Mobile Phone Shop   
3                Lounge                  Café    Mexican Restaurant   
4           Pizza Place           Coffee Shop                  Café   

  4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0           Yoga Studio        Supplement Shop            Steakhouse   
1   American Restaurant  Vietnamese Restaurant        Ice Cream Shop   
2         Grocery Store     Mexican Restaurant           Supermarket   
3         Deli / Bodega            Pizza Place    Chinese Restaurant   
4    Mexican Restaurant            Yoga Studio                Bakery   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                   Spa         Shopping Mall    Seafood Restaurant   
1       Bubble Tea Shop                   Spa   Dumpling Restaurant   
2        Sandwich Place    Chinese Restaurant      Tapas Restaurant   
3              Wine Bar   American Restaurant                Bakery   
4                  Park          Cocktail Bar        Sandwich Place   

  10th Most Common Venue  
0         Clothing Store  
1                 Bakery  
2            Coffee Shop  
3                   Park  
4                 School

In [32]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [33]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
2   Washington Heights                  Café                Bakery   
3               Inwood                Lounge                  Café   
4     Hamilton Heights           Pizza Place           Coffee Shop   
5       Manhattanville    Mexican Restaurant                  Park   
7          East Harlem    Mexican Restaurant                Bakery   
20     Lower East Side           Coffee Shop      Ramen Restaurant   
25    Manhattan Valley           Pizza Place           Coffee Shop   
36          Tudor City                  Park    Mexican Restaurant   

   3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
2      Mobile Phone Shop         Grocery Store         Mexican Restaurant   
3     Mexican Restaurant         Deli / Bodega                Pizza Place   
4                   Café    Mexican Restaurant                Yoga Studio   
5            Coffee Shop         Deli / Bodega         Italian Restaurant   
7          Deli / Bodega       Thai Restaurant  Latin American Restaurant   
20    Chinese Restaurant                  Café                Pizza Place   
25     Indian Restaurant           Yoga Studio            Thai Restaurant   
36                  Café      Greek Restaurant                Pizza Place   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2            Supermarket        Sandwich Place    Chinese Restaurant   
3     Chinese Restaurant              Wine Bar   American Restaurant   
4                 Bakery                  Park          Cocktail Bar   
5     Seafood Restaurant                 Diner            Bike Trail   
7     Chinese Restaurant          Liquor Store            Taco Place   
20           Art Gallery          Cocktail Bar        Sandwich Place   
25    Mexican Restaurant                  Café                   Bar   
36         Deli / Bodega                 Hotel                   Spa   

   9th Most Common Venue 10th Most Common Venue  
2       Tapas Restaurant            Coffee Shop  
3                 Bakery                   Park  
4         Sandwich Place                 School  
5                 Lounge       Sushi Restaurant  
7            Gas Station             Steakhouse  
20                Bakery    Japanese Restaurant  
25         Deli / Bodega             Playground  
36            Restaurant                  Diner

In [34]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1            Chinatown    Chinese Restaurant          Cocktail Bar   
6       Central Harlem    African Restaurant    Seafood Restaurant   
9            Yorkville                   Gym    Italian Restaurant   
10          Lenox Hill           Coffee Shop           Pizza Place   
11    Roosevelt Island        Sandwich Place                  Park   
12     Upper West Side    Italian Restaurant              Wine Bar   
16         Murray Hill           Coffee Shop        Sandwich Place   
19        East Village                   Bar              Wine Bar   
27            Gramercy                   Bar    Italian Restaurant   
29  Financial District           Coffee Shop            Steakhouse   
30       Carnegie Hill           Coffee Shop           Pizza Place   
33       Midtown South     Korean Restaurant                 Hotel   
34        Sutton Place  Gym / Fitness Center    Italian Restaurant   
35          Turtle Bay    Italian Restaurant            Steakhouse   

   3rd Most Common Venue   4th Most Common Venue     5th Most Common Venue  \
1     Salon / Barbershop     American Restaurant     Vietnamese Restaurant   
6    American Restaurant       French Restaurant                Public Art   
9            Coffee Shop                     Bar          Sushi Restaurant   
10    Italian Restaurant        Sushi Restaurant       Sporting Goods Shop   
11           Coffee Shop              Food Truck              Noodle House   
12                   Bar       Indian Restaurant  Mediterranean Restaurant   
16                 Hotel     Japanese Restaurant                       Bar   
19    Chinese Restaurant      Mexican Restaurant            Ice Cream Shop   
27            Bagel Shop     American Restaurant               Pizza Place   
29                 Hotel                     Gym                 Wine Shop   
30                  Café                     Bar               Yoga Studio   
33          Dessert Shop          Cosmetics Shop                 Hotel Bar   
34     Indian Restaurant  Furniture / Home Store       American Restaurant   
35           Coffee Shop        Sushi Restaurant          Ramen Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1         Ice Cream Shop       Bubble Tea Shop                   Spa   
6                    Bar    Chinese Restaurant          Dessert Shop   
9            Pizza Place             Wine Shop         Deli / Bodega   
10                   Gym  Gym / Fitness Center          Burger Joint   
11      Greek Restaurant          Liquor Store         Metro Station   
12                Bakery        Cosmetics Shop           Coffee Shop   
16    Italian Restaurant     French Restaurant  Gym / Fitness Center   
19           Pizza Place      Ramen Restaurant           Coffee Shop   
27       Thai Restaurant           Coffee Shop    Mexican Restaurant   
29                   Bar   American Restaurant           Event Space   
30             Bookstore   Japanese Restaurant         Grocery Store   
33   Japanese Restaurant           Coffee Shop   American Restaurant   
34                   Gym             Juice Bar          Dessert Shop   
35              Wine Bar     Indian Restaurant   Japanese Restaurant   

            9th Most Common Venue         10th Most Common Venue  
1             Dumpling Restaurant                         Bakery  
6                        Beer Bar                        Library  
9              Mexican Restaurant            Japanese Restaurant  
10                 Cosmetics Shop                  Deli / Bodega  
11                    Supermarket                Bubble Tea Shop  
12  Vegetarian / Vegan Restaurant                      Pet Store  
16                            Gym             Salon / Barbershop  
19                   Cocktail Bar  Vegetarian / Vegan Restaurant  
27                   Cocktail Bar         Thrift / Vintage Store  
29                    Pizza Place    

In [35]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
8       Upper East Side    Italian Restaurant               Exhibit   
13       Lincoln Square               Theater  Gym / Fitness Center   
14              Clinton               Theater  Gym / Fitness Center   
15              Midtown                 Hotel          Cocktail Bar   
17              Chelsea           Coffee Shop        Ice Cream Shop   
18    Greenwich Village    Italian Restaurant        Clothing Store   
21              Tribeca              Boutique                  Park   
22         Little Italy                Bakery                  Café   
23                 Soho        Clothing Store              Boutique   
24         West Village    Italian Restaurant        Cosmetics Shop   
26  Morningside Heights             Bookstore   American Restaurant   
28    Battery Park City                  Park           Coffee Shop   
31                 Noho    Italian Restaurant     French Restaurant   
32         Civic Center  Gym / Fitness Center     French Restaurant   
38             Flatiron                   Gym  Gym / Fitness Center   
39         Hudson Yards   American Restaurant  Gym / Fitness Center   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
8               Art Gallery           Coffee Shop  Gym / Fitness Center   
13                     Café          Concert Hall                 Plaza   
14      American Restaurant    Italian Restaurant                 Hotel   
15           Clothing Store               Theater           Coffee Shop   
17       Italian Restaurant             Nightclub                Bakery   
18         Sushi Restaurant     Indian Restaurant        Cosmetics Shop   
21       Italian Restaurant                  Café                   Spa   
22       Salon / Barbershop    Italian Restaurant        Sandwich Place   
23              Art Gallery         Women's Store            Shoe Store   
24  New American Restaurant                  Park              Wine Bar   
26                     Park           Coffee Shop            Food Truck   
28                    Hotel         Memorial Site                   Gym   
31             Cocktail Bar      Sushi Restaurant           Pizza Place   
32           Sandwich Place    Italian Restaurant                 Hotel   
38      Japanese Restaurant           Yoga Studio   American Restaurant   
39                     Café    Italian Restaurant                 Hotel   

   6th Most Common Venue  7th Most Common Venue     8th Most Common Venue  \
8                 Bakery      French Restaurant                       Spa   
13    Italian Restaurant  Performing Arts Venue       Indie Movie Theater   
14           Coffee Shop              Wine Shop                       Spa   
15                Bakery                    Spa       Japanese Restaurant   
17               Theater                  Hotel       American Restaurant   
18    Chinese Restaurant     Seafood Restaurant                      Café   
21   American Restaurant              Wine Shop                  Wine Bar   
22       Bubble Tea Shop         Clothing Store  Mediterranean Restaurant   
23           Men's Store     Italian Restaurant                     Hotel   
24   American Restaurant           Cocktail Bar                    Bakery   
26          Tennis Court          Deli / Bodega              Burger Joint   
28         Boat or Ferry              Wine Shop        Italian Restaurant   
31              Boutique            Coffee Shop                 Rock Club   
32           Coffee Shop            Yoga Studio       Sporting Goods Shop   
38                  Café     Salon / Barbershop   New American Restaurant   
39    Spanish Restaurant             Restaurant                   Theater   

   9th Most Common Venue    10th Most Common Venue  
8                  Hotel                 Juice Bar  
13           Opera House         French Restaurant  
14        Sandwich Place  Mediterranean Restaurant  
15   

In [36]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town                   Bar                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37            Playground           Pet Service           Gas Station   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37         Boat or Ferry        Farmers Market      Basketball Court   

   9th Most Common Venue 10th Most Common Venue  
37        Baseball Field        Harbor / Marina

In [37]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Marble Hill        Sandwich Place           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Discount Store           Yoga Studio       Supplement Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Steakhouse                   Spa         Shopping Mall   

  9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant         Clothing Store

## Toronto

In [38]:
!wget -q -O 'postalcodes_data.json' https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
print('Data downloaded!')

Data downloaded!


In [39]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [40]:
tables= pd.read_html(url)
tables[0]

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M5A  Downtown Toronto   
6        M6A        North York   
7        M6A        North York   
8        M7A      Queen's Park   
9        M8A      Not assigned   
10       M9A         Etobicoke   
11       M1B       Scarborough   
12       M1B       Scarborough   
13       M2B      Not assigned   
14       M3B        North York   
15       M4B         East York   
16       M4B         East York   
17       M5B  Downtown Toronto   
18       M5B  Downtown Toronto   
19       M6B        North York   
20       M7B      Not assigned   
21       M8B      Not assigned   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M9B         Etobicoke   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M1C       Scarborough   
30       M2C      Not assigned   
31       M3C        North York   
32       M3C        North York   
33       M4C         East York   
34       M5C  Downtown Toronto   
35       M6C              York   
36       M7C      Not assigned   
37       M8C      Not assigned   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M9C         Etobicoke   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M1E       Scarborough   
45       M2E      Not assigned   
46       M3E      Not assigned   
47       M4E      East Toronto   
48       M5E  Downtown Toronto   
49       M6E              York   
50       M7E      Not assigned   
51       M8E      Not assigned   
52       M9E      Not assigned   
53       M1G       Scarborough   
54       M2G      Not assigned   
55       M3G      Not assigned   
56       M4G         East York   
57       M5G  Downtown Toronto   
58       M6G  Downtown Toronto   
59       M7G      Not assigned   
60       M8G      Not assigned   
61       M9G      Not assigned   
62       M1H       Scarborough   
63       M2H        North York   
64       M3H        North York   
65       M3H        North York   
66       M3H        North York   
67       M4H         East York   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M5H  Downtown Toronto   
71       M6H      West Toronto   
72       M6H      West Toronto   
73       M7H      Not assigned   
74       M8H      Not assigned   
75       M9H      Not assigned   
76       M1J       Scarborough   
77       M2J        North York   
78       M2J        North York   
79       M2J        North York   
80       M3J        North York   
81       M3J        North York   
82       M4J         East York   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M5J  Downtown Toronto   
86       M6J      West Toronto   
87       M6J      West Toronto   
88       M7J      Not assigned   
89       M8J      Not assigned   
90       M9J      Not assigned   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M1K       Scarborough   
94       M2K        North York   
95       M3K        North York   
96       M3K        North York   
97       M4K      East Toronto   
98       M4K      East Toronto   
99       M5K  Downtown Toronto   
100      M5K  Downtown Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M6K      West Toronto   
104      M7K      Not assigned   
105      M8K      Not assigned   
106      M9K      Not assigned   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M1L       Scarborough   
110      M2L        North York   
111      M2L        North York   
112      M3L        North York   
113      M4L      East Toronto   
114      M4L      East Toronto   
115      M5L  Downtown Toronto   
116      M5L  Downtown Tor

In [43]:
df = pd.DataFrame(tables[0], 
               columns =['Postcode', 'Borough', 'Neighbourhood'])
dfassigned = df[df.Borough != 'Not assigned']

list = dfassigned.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join)
dfnew = pd.DataFrame(list).reset_index()
dfnew['Neighbourhood'] = np.where(dfnew['Neighbourhood'] == 'Not assigned', dfnew['Borough'], dfnew['Neighbourhood'])

In [44]:
dfnew.head()

Postcode      Borough                         Neighbourhood
0      M1B  Scarborough                         Rouge,Malvern
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2      M1E  Scarborough       Guildwood,Morningside,West Hill
3      M1G  Scarborough                                Woburn
4      M1H  Scarborough                             Cedarbrae

In [45]:
geo = pd.read_csv('http://cocl.us/Geospatial_data')

In [46]:
dfcombine = pd.concat([dfnew, geo], axis=1, sort=False)

In [47]:
dfcombine.drop(columns = ['Postal Code'])

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.23

In [48]:
address = 'Downtown Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6541737, -79.3808116451341.


In [49]:
# create map of Downtown Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(dfcombine['Latitude'], dfcombine['Longitude'], dfcombine['Borough'], dfcombine['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [50]:
toronto_data = dfcombine[dfcombine['Borough'].str.contains("Toronto")].reset_index(drop=True)

In [51]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [52]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [53]:
toronto_data.loc[0, 'Neighbourhood']

'The Beaches'

In [54]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [55]:
LIMIT = 100 
radius = 500 


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url 

'https://api.foursquare.com/v2/venues/explore?&client_id=MXQE1KN1ORGNBPLN03FEUOPV0SV24MS3F3NTZLNUD0XMOIGO&client_secret=L2WE2PO5U52P50DGVTWO5QF5BBT0AHJTI5NJK0XHQWYODNNT&v=20180605&ll=43.67635739999999,-79.2930312&radius=500&limit=100'

In [56]:
results = requests.get(url).json()

In [57]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0                   Glen Manor Ravine              Trail  43.676821 -79.293942
1  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
2                 Grover Pub and Grub                Pub  43.679181 -79.297215
3                       Upper Beaches       Neighborhood  43.680563 -79.292869

In [58]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


In [59]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvall

In [60]:
print(toronto_venues.shape)
toronto_venues.head()

(1701, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  The Beaches              43.676357              -79.293031   
1                  The Beaches              43.676357              -79.293031   
2                  The Beaches              43.676357              -79.293031   
3                  The Beaches              43.676357              -79.293031   
4  The Danforth West,Riverdale              43.679557              -79.352188   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            Pantheon       43.677621       -79.351434   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Greek Restaurant

In [61]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            57   
Brockton,Exhibition Place,Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                      19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     16   
Cabbagetown,St. James Town                                             43   
Central Bay Street                                                     82   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               16   
Church and Wellesley                                                   85   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             34   
Davisville North                                                        8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     15   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            21   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          33   
Harbourfront East,Toronto Islands,Union Station                       100   
Harbourfront,Regent Park                                               49   
High Park,The Junction South                                           24   
Lawrence Park                                                           4   
Little Portugal,Trinity                                                63   
Moore Park,Summerhill East                                              1   
North Toronto West                                                     17   
Parkdale,Roncesvalles                                                  15   
Rosedale                                                                5   
Roselawn                                                                1   
Runnymede,Swansea                                                      40   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        95   
Studio District                                                        41   
The Annex,North Midtown,Yorkville                                      24   
The Beaches                                                             4   
The Beaches West,India Bazaar                                          21   
The Danforth West,Riverdale                                            42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             57   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      16   
Cabbagetown,St. James Town                                              43   
Central Bay Street                                                      82   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                16   
Church and Wellesley                                                    85   
Commerce Court,Victoria Hotel      

In [62]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 239 uniques categories.


In [65]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

In [66]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.011765   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19    Harbourfront East,Toronto Islands,Union Station           0.000000   
20                           Harbourfront,Regent Park           0.000000   
21                       High Park,The Junction South           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                           Rosedale           0.000000   
28                                           Roselawn           0.000000   
29                                  Runnymede,Swansea           0.000000   
30                            Ryerson,Garden District           0.000000   
31                                     St. James Town           0.000000   
32                    Stn A PO Boxes 25 The Esplanade           0.000000   
33                                    Studio District           0.000000   
34                  The Annex,North Midtown,Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                      The Beaches West,India Bazaar           0.000000   
37                        The Danforth West,Riverdale           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0000              0.0000        0.0000           0.000   
4    0.0625              0.0625        0.0625           0.125   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000              0.0000        0.0000           0.000   
15   0.0000              0.0000     

In [67]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [68]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Café                   Bar             Steakhouse   
1                Bakery          Cocktail Bar               Beer Bar   
2        Breakfast Spot                  Café  Performing Arts Venue   
3           Yoga Studio                   Spa          Garden Center   
4        Airport Lounge      Airport Terminal               Boutique   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant        Breakfast Spot      Asian Restaurant   
1                  Café            Steakhouse    Seafood Restaurant   
2         Burrito Place               Stadium  Caribbean Restaurant   
3                Garden  Fast Food Restaurant        Farmers Market   
4               Airport    Airport Food Court          Airport Gate   

  8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0        Cosmetics Shop                 Hotel              Restaurant  
1           Cheese Shop        Farmers Market                  Bistro  
2            Restaurant                   Bar  Furniture / Home Store  
3                  Park            Comic Shop             Pizza Place  
4                   Bar       Harbor / Marina        Sculpture Garden

In [69]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [75]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels4', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

In [76]:
toronto_merged.drop(columns=['Cluster Labels3'])

Postcode           Borough  \
0       M4E      East Toronto   
1       M4K      East Toronto   
2       M4L      East Toronto   
3       M4M      East Toronto   
4       M4N   Central Toronto   
5       M4P   Central Toronto   
6       M4R   Central Toronto   
7       M4S   Central Toronto   
8       M4T   Central Toronto   
9       M4V   Central Toronto   
10      M4W  Downtown Toronto   
11      M4X  Downtown Toronto   
12      M4Y  Downtown Toronto   
13      M5A  Downtown Toronto   
14      M5B  Downtown Toronto   
15      M5C  Downtown Toronto   
16      M5E  Downtown Toronto   
17      M5G  Downtown Toronto   
18      M5H  Downtown Toronto   
19      M5J  Downtown Toronto   
20      M5K  Downtown Toronto   
21      M5L  Downtown Toronto   
22      M5N   Central Toronto   
23      M5P   Central Toronto   
24      M5R   Central Toronto   
25      M5S  Downtown Toronto   
26      M5T  Downtown Toronto   
27      M5V  Downtown Toronto   
28      M5W  Downtown Toronto   
29      M5X  Downtown Toronto   
30      M6G  Downtown Toronto   
31      M6H      West Toronto   
32      M6J      West Toronto   
33      M6K      West Toronto   
34      M6P      West Toronto   
35      M6R      West Toronto   
36      M6S      West Toronto   
37      M7Y      East Toronto   

                                        Neighbourhood Postal Code   Latitude  \
0                                         The Beaches         M4E  43.676357   
1                         The Danforth West,Riverdale         M4K  43.679557   
2                       The Beaches West,India Bazaar         M4L  43.668999   
3                                     Studio District         M4M  43.659526   
4                                       Lawrence Park         M4N  43.728020   
5                                    Davisville North         M4P  43.712751   
6                                  North Toronto West         M4R  43.715383   
7                                          Davisville         M4S  43.704324   
8                          Moore Park,Summerhill East         M4T  43.689574   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...         M4V  43.686412   
10                                           Rosedale         M4W  43.679563   
11                         Cabbagetown,St. James Town         M4X  43.667967   
12                               Church and Wellesley         M4Y  43.665860   
13                           Harbourfront,Regent Park         M5A  43.654260   
14                            Ryerson,Garden District         M5B  43.657162   
15                                     St. James Town         M5C  43.651494   
16                                        Berczy Park         M5E  43.644771   
17                                 Central Bay Street         M5G  43.657952   
18                             Adelaide,King,Richmond         M5H  43.650571   
19    Harbourfront East,Toronto Islands,Union Station         M5J  43.640816   
20            Design Exchange,Toronto Dominion Centre         M5K  43.647177   
21                      Commerce Court,Victoria Hotel         M5L  43.648198   
22                                           Roselawn         M5N  43.711695   
23                 Forest Hill North,Forest Hill West         M5P  43.696948   
24                  The Annex,North Midtown,Yorkville         M5R  43.672710   
25                      Harbord,University of Toronto         M5S  43.662696   
26            Chinatown,Grange Park,Kensington Market         M5T  43.653206   
27  CN Tower,Bathurst Quay,Island airport,Harbourf...         M5V  43.628947   
28                    Stn A PO Boxes 25 The Esplanade         M5W  43.646435   
29              First Canadian Place,Underground city         M5X  43.648429   
30                                           Christie         M6G  43.669542   
31                        Dovercourt Village,Dufferin         M6H  43.669005   
32                            Little Portugal,Trinity         M6J  43.64792

In [78]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels4']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Results

NYC has 26 venues while Toronto only has 4 venues.  NYC has 342 uniques venue categories while Toronto has 238 uniques venue categories. Both cities have many coffe shops and restaurants.   
From both maps and dataframes we can see that, New York City is more crowded and focused, with more venues and neighborhoods. It's also more diverse in terms of categories.  
However, both cities share the similar patterns of neighborhoods, including park, coffe shop, restaurant etc. They are typical financial centers in their districts.

## Discussion

Based on the analysis we did above, I recommend companies who are looking for new locations know more about NYC and Toronto.   
Both cities have high volumn of population, diverse kinds of people and industries. Financial and Service companies should especially consider these two cities.  
However, this analysis didn't explore details of the neighborhoods, further research can be conducted.

## Conclusion

In conclusion, this is the complete analysis I made to compare the neighborhoods of New York City and Toronto, two major cities of its respective countries.  